File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    List<String> res = new ArrayList<>();
    String allDigits;
    HashMap<Character, String> map = new HashMap<>();
    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0) return res;
        allDigits = digits;
        map.put('2', "abc");
        map.put('3', "def");
        map.put('4', "ghi");
        map.put('5', "jkl");
        map.put('6', "mno");
        map.put('7', "pqrs");
        map.put('8', "tuv");
        map.put('9', "wxyz");
        backtrack(0, new StringBuilder());
        return res;
    }
    
    public void backtrack(int idx, StringBuilder curr) {
        if (idx == allDigits.length()) {
            res.add(curr.toString());
            return;
        }
        String possibilities = map.get(allDigits.charAt(idx));
        for (char possibility: possibilities.toCharArray()) {
            curr.append(possibility);
            backtrack(idx + 1, curr);
            curr.deleteCharAt(curr.length() - 1);
        }
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
         
        unordered_map<char, vector<char>> map;
        map['2'] = {'a', 'b', 'c'};
        map['3'] = {'d', 'e', 'f'};
        map['4'] = {'g', 'h', 'i'};
        map['5'] = {'j', 'k', 'l'};
        map['6'] = {'m', 'n', 'o'};
        map['7'] = {'p', 'q', 'r', 's'};
        map['8'] = {'t', 'u', 'v'};
        map['9'] = {'w', 'x', 'y', 'z'};
        vector<string> res;
        if (digits.length() == 0) return res;
        string curr;
        backtrack(0, digits, curr, res, map);
        return res;
    }
    
    void backtrack(int i, string& digits, string& curr, vector<string>& res, unordered_map<char, vector<char>>& map) {
        if (i == digits.length()) {
            res.push_back(curr);
            return;
        }
        
        vector<char> choices = map[digits[i]];
        for (char choice: choices) {
            curr += choice;
            backtrack(i + 1, digits, curr, res, map);
            curr.pop_back();
        }
    }
};